rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1228 @@
+
+### Formatting gotchas requiring manual intervention
+
+Some code patterns in Nixpkgs will result in a sub-optimal format,
+because an auto-formatter cannot do exceptions based on context.
+A lot of the times though, the same program can be equivalently expressed in a prettier way.
+
+#### CLI argument flags
+
+It is common to pass in CLI flags (e.g. to builders) like this:
+
+```nix
+[
+  "--some-flag" "some-value"
+]
+```
+
+However, this will be formatted sub-optimally:
+
+```nix
+[
+  "--some-flag"
+  "some-value"
+]
+```
+
+If the CLI also accepts GNU-style flags, a more structured helper can be used instead:
+
+```nix
+lib.cli.toGNUCommandLine { } {
+  some-flag = "some-value";
+}
+```
+
+#### Singleton lists
+
+Sometimes a list only needs a single element
+and there's no expectation to add more in the future.
+This would be formatted like this:
+
+```nix
+{
+  list = [
+    {
+      foo = 10;
+      bar = 20;
+    }
+  ];
+}
+```
+
+In this case one level of indentation can be saved using [`lib.singleton`](https://nixos.org/manual/nixpkgs/stable#function-library-lib.lists.singleton):
+
+```nix
+{
+  list = lib.singleton {
+    foo = 10;
+    bar = 20;
+  };
+}
+```
+
+## Drawbacks
+
+- No automatic code format can be as pretty as a carefully crafted manual one. There will always be "ugly" edge cases.
+  - However, we argue that on average the new format will be an improvement,
+    and that contributors should not waste their time making the code slightly more pretty.
+- Every formatter will have bugs, but the Nix format team will be able to make fixes.
+- Having a commit that changes the formatting, can make git blame harder to use.
+  - This can be [worked around in interfaces](#git-blames).
+
+## Alternatives
+
+- Keep the status quo of not having an official formatter.
+  The danger is that this creates discord within the Nix community.
+  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
+- Specify a different format.
+- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
+- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
+  However, such an approach would not reduce the amount of merge conflicts,
+  and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+- Should the line length limit be specified?
+
+## Future work
+
+- General style guidelines beyond AST reformatting
+- Making widespread use of linters like Nixpkgs-hammering
+- Enforcing the format to other official repositories
+
+-----
+
+## Initial standard Nix format
+
+Terms and definitions:
+- Brackets: `[]`
+- Braces: `{}`
+- Parentheses: `()`
+
+- Line breaks may not be preserved, but empty lines must always be preserved.
```

Comment:
More ‘may not’ pedantry:

```suggestion
- Line breaks may be added or removed, but empty lines must not be added, and empty lines must be removed if they follow another empty line (and must not be removed otherwise).
```

Also, while I'm picking this nit: would we want this to say that exactly one line break (i.e., no empty lines) must appear at the end of a file?

Reply:
EOL on EOF.

Reply:
> Also, while I'm picking this nit: would we want this to say that exactly one line break (i.e., no empty lines) must appear at the end of a file?

We just added an editorconfig yesterday:
https://github.com/nix-rfc-101/rfcs/commit/5fd936505c25533875cb23ae32ce69a018516e09

Reply:
@rhendric Regarding the wording on line breaks and various other improvements, see https://github.com/nix-rfc-101/rfcs/pull/15. In particular I'm changing the wording to this there:

> Line breaks may be added or removed. Consecutive empty lines must be turned into a single empty line.